### PR TITLE
[trainer] feat: add mindspeedllm backend engine support on NPU.

### PR DESCRIPTION
**Pull request by pengnuoheng**
### What does this PR do?

Add mindspeedllm backend engine support on NPU.

### Checklist Before Starting

- [ ] Search for similar PRs. Paste at least one query link here: ...
- [ ] Format the PR title as `[{modules}] {type}: {description}` (This will be checked by the CI)
  - `{modules}` include `fsdp`, `megatron`, `veomni`, `sglang`, `vllm`, `rollout`, `trainer`, `ci`, `training_utils`, `recipe`, `hardware`, `deployment`, `ray`, `worker`, `single_controller`, `misc`, `perf`, `model`, `algo`, `env`, `tool`, `ckpt`, `doc`, `data`, `cfg`, `reward`, `fully_async`, `one_step_off`
  - If this PR involves multiple modules, separate them with `,` like `[megatron, fsdp, doc]`
  - `{type}` is in `feat`, `fix`, `refactor`, `chore`, `test`
  - If this PR breaks any API (CLI arguments, config, function signature, etc.), add `[BREAKING]` to the beginning of the title.
  - Example: `[BREAKING][fsdp, megatron] feat: dynamic batching`

### Test

> For changes that can not be tested by CI (e.g., algorithm implementation, new model support), validate by experiment(s) and show results like training curve plots, evaluation results, etc.

### API and Usage Example

> Demonstrate how the API changes if any, and provide usage example(s) if possible.

```python
# Add code snippet or script demonstrating how to use this
```

### Design & Code Changes

RFC:https://github.com/verl-project/verl/discussions/5063

### Checklist Before Submitting

> [!IMPORTANT]
> Please check all the following items before requesting a review, otherwise the reviewer might deprioritize this PR for review.

- [ ] Read the [Contribute Guide](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md).
- [ ] Apply [pre-commit checks](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md#code-linting-and-formatting): `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always`
- [ ] Add / Update [the documentation](https://github.com/volcengine/verl/tree/main/docs).
- [ ] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/volcengine/verl/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: ...
- [ ] Once your PR is ready for CI, send a message in [the `ci-request` channel](https://verl-project.slack.com/archives/C091TCESWB1) in [the `verl` Slack workspace](https://join.slack.com/t/verl-project/shared_invite/zt-3855yhg8g-CTkqXu~hKojPCmo7k_yXTQ). (If not accessible, please try [the Feishu group (飞书群)](https://applink.larkoffice.com/client/chat/chatter/add_by_link?link_token=772jd4f1-cd91-441e-a820-498c6614126a).)
- [ ] If your PR is related to the `recipe` submodule, please also update the reference to the submodule commit via `git submodule update --remote` or `cd recipe && git pull origin main`.

### CODE REVIEW

File: verl/models/mcore/util.py [hunk 1]
```diff
@@ -319,6 +319,23 @@ def preprocess_thd_no_padding(
...
+        set_actual_seq_len(cu_seqlens_padded)
+        # Generate position IDs within each padded segment
+        pack_length = int(seqlens_in_batch_padded.sum().item())
+        position_ids_packed = torch.zeros(pack_length, dtype=torch.int32, device=input_ids.device)
+        for i in range(batch_size):
+            start = cu_seqlens_padded[i].item()
+            end = cu_seqlens_padded[i + 1].item()
+            position_ids_packed[start:end] = torch.arange(end - start, dtype=torch.int32, device=input_ids.device)
+
+        set_position_ids(position_ids_packed.unsqueeze(0).transpose(0, 1).contiguous())
```
Comment (psyloy) [inline verl/models/mcore/util.py]: "Is this logic only used by the MindSpeed-LLM backend? Will it affect other training backends?

File: verl/utils/vllm/npu_vllm_patch.py
Comment (psyloy) [inline verl/utils/vllm/npu_vllm_patch.py:183]: Does this file not involve any modifications?

File: verl/workers/config/engine.py [hunk 2]
```diff
@@ -518,6 +519,30 @@ def __post_init__(self):
...
+        mm_kwargs (str): mindspeedmm engine kwargs.
+    """
+
+    strategy: str = "mindspeedllm"
+    llm_kwargs: dict[str, Any] = field(default_factory=dict)
+    mm_kwargs: dict[str, Any] = field(default_factory=dict)
+
+    def __post_init__(self) -> None:
+        """config validation logics go here"""
+        assert self.strategy in ["mindspeedllm", "mindspeedmm"], f"strategy {self.strategy} not supported"
```
Comment (psyloy) [inline verl/workers/config/engine.py]: "Would it be better to write 'mindspeedllm' and 'mindspeedmm' as 'mindspeed_llm' and 'mindspeed_mm' instead?"

File: verl/utils/vllm/npu_vllm_patch.py
Comment (pengnuoheng) [reply verl/utils/vllm/npu_vllm_patch.py:183]: This file modifies the code in order to pass ut

File: verl/workers/config/engine.py (hunk 2 above)
Comment (pengnuoheng) [reply verl/workers/config/engine.py]: already modified

File: verl/models/mcore/util.py (hunk 1 above)
Comment (pengnuoheng) [reply verl/models/mcore/util.py]: it's also used by MindSpeed backend, and verified it will not affect the effect of MindSpeed backend.

File: docs/ascend_tutorial/quick_start/ascend_quick_start.rst [hunk 3]
```diff
@@ -139,6 +139,36 @@ MindSpeed 对应 Megatron-LM 后端使用场景，使用方式如下：
     3. 更多特性信息可参考 `MindSpeed & verl 文档 <https://gitcode.com/Ascend/MindSpeed/blob/master/docs/user-guide/verl.md>`_ 。
 
 
+新增 MindSpeed-LLM 训练后端支持
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+如需使能 MindSpeed-LLM 训练后端，需要额外下载 MindSpeed-LLM 。
+需要注意的是，MindSpeed-LLM 训练后端依赖 MindSpeed-LLM master 分支、 MindSpeed master 分支以及 Megatron-LM core_v0.12.1 分支。
```
Comment (wucong25) [inline docs/ascend_tutorial/quick_start/ascend_quick_start.rst:145]: 新增LLM后端后，需要对当前后端的功能进行界限，建议列一个模型列表，表明LLM可以跑哪些模型

File: .github/workflows/e2e_ascend.yml [hunk 4]
```diff
@@ -119,6 +119,60 @@ jobs:
...
+    name: E2E Ascend testing for RL training scenarios of LLM models using MindSpeed_LLM engine
+    runs-on: linux-aarch64-a2b3-8
+    timeout-minutes: 120
+    container:
+      image: swr.cn-southwest-2.myhuaweicloud.com/modelfoundry/ascend-ci/verl/verl:verl-sglang-8.3.rc1-910b-ubuntu22.04-py3.11-latest
+      options: >-
+        --shm-size 16g
+    env:
+      HF_ENDPOINT: "https://hf-mirror.com"
+      HF_HUB_ENABLE_HF_TRANSFER: "0" # This is more stable
```
Comment (wucong25) [inline .github/workflows/e2e_ascend.yml:133]: 后续可以上一个精度看护nightly ci

**Comment by hustmf** [comment]
validation result
image
image

File: verl/trainer/config/ppo_mindspeed_trainer.yaml [hunk 5]
```diff
@@ -0,0 +1,244 @@
+# specify the default per-component configs
+defaults:
```
Comment (wuxibin89) [inline verl/trainer/config/ppo_mindspeed_trainer.yaml]: Please unify this config to `verl/trainer/config/ppo_trainer.yaml`
https://github.com/verl-project/verl/blob/main/verl/trainer/config/ppo_megatron_trainer.yaml

File: verl/models/mcore/util.py [hunk 6]
```diff
@@ -333,6 +333,23 @@ def preprocess_thd_engine(
...
+    except ImportError as e:
+        logger.warning(f"Could not import mindspeed modules, skipping position_id setting: {e}")
```
Comment (wuxibin89) [inline verl/models/mcore/util.py]: This warning is confused in GPU environment.

File: verl/models/mcore/util.py [hunk 7]
```diff
@@ -333,6 +333,23 @@ def preprocess_thd_engine(
     cu_seqlens_padded = torch.zeros(batch_size + 1, dtype=torch.int32, device=input_ids.device)
     cu_seqlens_padded[1:] = torch.cumsum(seqlens_in_batch_padded, dim=0)
 
+    try:
```
Comment (wuxibin89) [inline verl/models/mcore/util.py]: wrap this to a separate function

File: verl/trainer/config/engine/mindspeed.yaml [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+  group_query_attention: true
+
+  transformer_impl: transformer_engine
+
+  no_pad_to_seq_lengths: true
+  reset_attention_mask: true
+  context_parallel_algo: ulysses_cp_algo
+  attention_mask_type: general
+
+  #optimizer
```
Comment (psyloy) [inline verl/trainer/config/engine/mindspeed.yaml]: 这块注释可以参考mindspeed_critic文件写详细一些

File: verl/trainer/config/critic/mindspeed_critic.yaml [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+      - mindspeed
+
+  # dp actor config, inheriting from trainer/config/critic/critic.yaml
+  - critic
+
+  # load the reference default config, then apply the fields in the current yaml
+  - _self_
+
+# Required when using verl.utils.omega_conf_to_dataclass to instantiate dataclass configs
+_target_: verl.workers.config.MindSpeedCriticConfig
```
Comment (wuxibin89) [inline verl/trainer/config/critic/mindspeed_critic.yaml:19]: 简化这个文件，参考其他backend，model不需要单独定义
